qutip/tests/core/test_qobj.py

```diff
@@ -76,6 +76,38 @@ def test_QobjData():
     assert isinstance(q2.data, qutip.core.data.Data)
 
 
+@pytest.mark.parametrize("data",
+                         [
+                            qutip.data.Dense(_random_not_singular(2)),
+                            qutip.data.csr.identity(2),
+                            qutip.Qobj(_random_not_singular(2)),
+                            _random_not_singular(2),
+                         ],
+                         ids=[
+                             "Dense",
+                             "CSR",
+                             "Qobj",
+                             "Numpy",
+                         ])
+@pytest.mark.parametrize("copy", [True, False])
+def test_QobjCopyArgument(data, copy):
+    """Tests that Qobj copy argument works properly when instantiating Qobj."""
+    qobj_data = qutip.Qobj(data, copy=copy).data
+
+    if isinstance(data, qutip.Qobj):
+        data = data.data  # Qobj copies the data of another Qobj
+
+    # For numpy object we compare with data's data. This should be dense so we
+    # get its data as ndarray.
+    if isinstance(data, np.ndarray):
+        qobj_data = qobj_data.as_ndarray()
```

Comment:
This does not seem to work. I am not sure when it should work or how to make it work. What I was expecting is that if the `ndarray` is passed with `copy=False` then I should be able to retrieve it somehow from `qobj.data`. I guess I could check that operating in `qobj` does not change input `data`? Any ideas of how this should be done @jakelishman?

Reply:
`copy=False` refers to the data, not the entire object.  This allows us to avoid copies even when the shape is being changed (say you started with a 1D array for a ket, or whatever). You can check that the data wasn't copied here with `np.shares_memory`:
```python
>>> x = np.arange(5, dtype=complex)
>>> # This is explicitly a no-op in terms of data, but creates a new Python object.
>>> y = x.reshape(x.shape, order='A')
>>> x is y
False
>>> # But they still point to the same place in memory.
>>> np.shares_memory(x, y)
True
```

Similar things happen inside `data.Dense` in a few places, to allow the initialiser to do its job.

Reply:
Thanks @jakelishman for the clear explanation! It now works!